sklearn/_fast_gradient_boosting/loss.py

```diff
@@ -0,0 +1,238 @@
+            If hessians are constant (e.g. for ``LeastSquares`` loss, the
+            array is initialized to ``1``.
+        """
+        shape = (prediction_dim, n_samples)
+        gradients = np.empty(shape=shape, dtype=G_H_DTYPE)
+        if self.hessians_are_constant:
+            # if the hessians are constant, we consider they are equal to 1.
+            # this is correct as long as we adjust the gradients. See e.g. LS
+            # loss
+            hessians = np.ones(shape=(1, 1), dtype=G_H_DTYPE)
+        else:
+            hessians = np.empty(shape=shape, dtype=G_H_DTYPE)
+
+        return gradients, hessians
+
+    @abstractmethod
+    def get_baseline_prediction(self, y_train, prediction_dim):
+        """Return initial predictions (before the first iteration).
+
+        Parameters
+        ----------
+        y_train : array-like, shape=(n_samples,)
+            The target training values.
+        prediction_dim : int
+            The dimension of one prediction: 1 for binary classification and
+            regression, n_classes for multiclass classification.
+
+        Returns
+        -------
+        baseline_prediction: float or array of shape (1, prediction_dim)
+            The baseline prediction.
+        """
+        pass
+
+    @abstractmethod
+    def update_gradients_and_hessians(self, gradients, hessians, y_true,
+                                      raw_predictions):
+        """Update gradients and hessians arrays, inplace.
+
+        The gradients (resp. hessians) are the first (resp. second) order
+        derivatives of the loss for each sample with respect to the
+        predictions of model, evaluated at iteration ``i - 1``.
+
+        Parameters
+        ----------
+        gradients : array-like, shape=(prediction_dim, n_samples)
+            The gradients (treated as OUT array).
+        hessians : array-like, shape=(prediction_dim, n_samples) or \
+            (1,)
+            The hessians (treated as OUT array).
+        y_true : array-like, shape=(n_samples,)
+            The true target values or each training sample.
+        raw_predictions : array-like, shape=(prediction_dim, n_samples)
+            The raw_predictions (i.e. values from the trees) of the tree
+            ensemble at iteration ``i - 1``.
+        """
+        pass
+
+
+class LeastSquares(BaseLoss):
+    """Least squares loss, for regression.
+
+    For a given sample x_i, least squares loss is defined as::
+
+        loss(x_i) = (y_true_i - raw_pred_i)**2
+    """
+
+    hessians_are_constant = True
+
+    def __call__(self, y_true, raw_predictions, average=True):
+        # shape (1, n_samples) --> (n_samples,). reshape(-1) is more likely to
+        # return a view.
+        raw_predictions = raw_predictions.reshape(-1)
+        loss = np.power(y_true - raw_predictions, 2)
+        return loss.mean() if average else loss
+
+    def get_baseline_prediction(self, y_train, prediction_dim):
```

Comment:
shouldn't all public methods have a docstring? I'm not sure what `prediction_dim`'s supposed to do for instance.

Reply:
I commented the methods in the base class, thinking it would be enough (all these classes are private anyway). Should I also add the docstring to the child classe?

Reply:
no it's fine then, somehow even when reading from top to bottom, I missed that, my bad.